## RNA-Seq analysis workflow using High Performance Computing

Welcome to the overview of the course
*'RNA-Seq analysis workflow using High Performance Computing'*.

This course will be delivered online over two half-days:  

**Part 1: 09.30 - 12.30 9th November** \
**Part 2: 09.30 - 12.30 12th November**

Organiser: [**Dr Prasanth Sivakumar**](https://scholar.google.co.uk/citations?user=kggRsMsAAAAJ&hl=en) \
Contact: prasanth.sivakumar@kcl.ac.uk \
Bioinformatician with the
NIHR Biomedical Research Centre
Guy's and St Thomas' NHS Foundation Trust and King's College London
16th Floor Tower Wing
Guy's Hospital
Great Maze Pond, London SE1 9RT


### General Information: 
RNA sequencing is a popular next generation sequencing method that enables the identification and quantification of both known and novel transcripts across the entire transcriptome, but requires several complex and computationally-intensive processing steps. This course gives an introduction into how researchers with a beginner's understanding on the command line interface can gain experience with such methods on Rosalind, KCL's high performance computing system. RNA-Seq data will be processed and analysed hands-on by working through a guided example using a sample of real published datasets. 

### Pre-requisite:

* Basic knowledge of working on the command line is required for Part 1 of the course. 
* Basic knowledge of R is required for Part 2 of the course. 
* An account on Rosalind is necessary and access to your home directory must be fully set up prior to attending this course. Request an account [here.](https://rosalind.kcl.ac.uk/requesting_access/) 
* Basic understanding of RNA sequencing is not necessary but would be beneficial. 
* To save 

### Format:
Students will use their own laptops. This will allow them to store notes and scripts, as well as enabling them to carry out hands-on analysis. They will use their own Rosalind accounts to follow a guided workflow and obtain results from a sample of a published dataset.

### Learning Outcomes:
Students will:

*	Gain an overview of the steps involved in RNA-Seq data processing and analysis, from QC and alignment through to producing a count matrix, identifying differentially expressed genes, and interpreting results.
*	Learn the methods involved in submitting scripts and running jobs in a high performance computing system.
*	Be able to perform a basic RNA-Seq data processing workflow from raw data files through to outputs of differential gene expression.
*	Understand how to use RNA-Seq data and differential gene expression to identify results of interest based on their biological question.
*	Improve on their ability to utilise the command line.
*	By the end of the course, successfully obtain a fully-processed differential gene expression dataset.

### Collaborations
There are numerous downstream analysis options available given the quality checked and cleaned differential gene expression data. These approaches must be tailored to requirements in a project-specific manner. If you are interested in conducting such analyses or potential bioinformatics collaborations, please contact **Dr. Mansoor Saqi**, Head of Translational Bioinformatics at the NIHR Biomedical Research Centre, at mansoor.saqi@kcl.ac.uk


